pycparser/c_parser.py

```diff
@@ -973,47 +1011,43 @@ def p_enumerator(self, p):
 
         p[0] = enumerator
 
-    def p_declarator_1(self, p):
-        """ declarator  : direct_declarator
+    def p_declarator(self, p):
+        """ declarator  : id_declarator
+                        | typeid_declarator
         """
         p[0] = p[1]
 
-    def p_declarator_2(self, p):
-        """ declarator  : pointer direct_declarator
+    @parameterized(('id', 'ID'), ('typeid', 'TYPEID'), ('typeid_noparen', 'TYPEID'))
+    def _p_XXX_declarator_1(self, p):
```

Comment:
What do you mean by "special name", the XXX part or the leading underscore? If you're talking about the XXX, I had assumed that the `p_` functions had to be named the same as their nonterminal. If that's not the case, I can remove the XXX.

If you're talking about the underscore, it's required so the function doesn't get treated as a rule directly. If you omitted the underscore, you'd have to either rename or delete the template method to keep this from happening, which means you'd need different behavior when first instantiating a CParser.

This arises because I wrote `_create_param_rules()` in a similar way as `create_opt_rule()`, which is called during `__init__()`. What makes more sense to me is to instead use a very simple metaclass so this modification (or creation) of methods only happens at class-creation time, not every time a parser is instantiated. The metaclass's `__new__()` method simply loops through the class members, modifies them as described, and passes the modified classdict to `type()`.